Enhancements to configuration: Grouped configuration settings & custom bake paths

jamisonbryant opened the pull request:
References #178 

My app wants to bake the fixture factories into a separate directory than the default, only I found that that was not possible in the current plugin. This includes a patch to fix that (with tests). Users can now set `testFixtureOutputDir` to override the output directory where factories get baked.

**Secondarily, and very importantly,** all configuration settings for the plugin have been _namespaced_. This is something I brought up in my original feature ticket (linked above) and no one seemed to have a problem with it so I went for it. It is common practice for plugins to keep their configuration settings out of the global namespace, and so I have moved all configuration settings for this plugin to the `FixtureFactories` namespace:

```php
[
    'FixtureFactories' => [
        'testFixtureGlobalBehaviors' => [...],
        'testFixtureOutputDir' => 'my/output/dir',
    ],
]
```

The latter is **not** a backwards compatible change, so beware.

jamisonbryant commented on docs/setup.md:
```diff
@@ -24,6 +24,9 @@ protected function bootstrapCli(): void
 
+## CakePHP 4.4+
+For CakePHP 4.4 or higher applications, add the `TruncateDirtyTables` trait to the top of your test classes to automatically truncate dirty tables after each test runs.
+No modification to your PHPUnit configuration is required when using the trait.
```
as soon as #181 is merged I will rebase this branch and this will go away

pabloelcolombiano replied:
Note that this should read CakePHP ≥ 4.3

jamisonbryant replied:
Updates made to file in other PR

jamisonbryant commented on src/Command/BakeFixtureFactoryCommand.php:
```diff
@@ -110,10 +110,13 @@ public function setTable(string $tableName, ConsoleIo $io)
      */
     public function getPath(Arguments $args): string
```
I don't understand what the purpose of the `$args` parameter? It does not appear to be referenced inside the function. Can it be removed or does it have a purpose I'm not aware of?

pabloelcolombiano replied:
See the signature of `parent::getPath()`

jamisonbryant commented on tests/TestCase/Command/BakeFixtureFactoryCommandTest.php:
```diff
@@ -63,6 +65,18 @@ public function testGetFileName()
 
+    public function testGetPath()
+    {
+        $args = new Arguments([], [], []);
```
This variable is required to pass to `getPath()` even though the function doesn't use that parameter, so this is why I'm wondering if it can be removed? Thanks!

pabloelcolombiano replied:
See the comment above, $args should remain

jamisonbryant commented:
The three PHPStan errors are also showing up on the `next` branch, they were not introduced in this branch.

I believe commands are supposed to `return $this->abort('msg')` instead of calling the function and then returning something else, which is what it looks like `BakeFixtureFactory` and `Setup` are doing. 

I think these two issues should possibly be addressed in a separate PR, thoughts?

jamisonbryant commented:
@pabloelcolombiano I have made the requested changes, and also fixed the PHPStan errors. All tests are still passing even though I removed some `returns` (they weren't getting hit anyway because `abort()` throws a `StopException`).

I noticed you also have Psalm configured, that is currently _not_ passing although it's because it is not inferring the same return type on two methods in the `BaseFactory`. I feel like its inferred return type is less detailed than the documented one, and so no change is necessary. More detail better than less. Let me know if you think different.

Ready for re-review.
